autotag.go

```diff
@@ -58,56 +55,64 @@ func NewRepo(repoPath string) (*GitRepo, error) {
 	return r, nil
 }
 
+// Parse tags on repo and
 func (r *GitRepo) parseTags() error {
-	log.Print("Parsing repo tags")
-	iter, err := r.Repo.NewReferenceIterator()
+	log.Println("Parsing repository tags")
+
+	versions := make(map[*version.Version]*git.Tag)
+
+	tags, err := r.Repo.GetTags()
 	if err != nil {
-		return fmt.Errorf("ouch, couldn't get a ref on repo this shouldn't happen ever: %s", err)
+		return fmt.Errorf("failed to fetch tags: %s", err.Error())
 	}
 
-	var versions []*version.Version
-	var lastParsedTagRef string
+	for _, tag := range tags {
+		if v, err := maybeVersionFromTag(tag); err == nil {
+
+			t, err := r.Repo.GetTag(tag)
+			if err != nil {
+				log.Printf("Error fetching commit for tag '%s' %s", tag, err)
+				continue
+			}
+
+			//			t := r.Repo.GetTag(tag)
+			//			t.Commit().
 
-	// step through the refrences
-	for ref, err := iter.Next(); err == nil; ref, err = iter.Next() {
-		log.Println("checking ref: ", ref.Name())
-		if v, err := maybeVersionFromTag(ref); err == nil {
-			versions = append(versions, v)
-			lastParsedTagRef = ref.Target().String()
+			//	log.Printf("Found tag %s ref: %s", v.String(), commit.Id)
+			versions[v] = t
 		}
 	}
 
-	cRange := fmt.Sprintf("%s..HEAD", lastParsedTagRef)
-	//	log.Println("Check range:", cRange)
-	if wErr := r.walker.PushRange(cRange); wErr != nil {
-		log.Fatal(wErr)
+	keys := make([]*version.Version, 0, len(versions))
+	for key := range versions {
+		keys = append(keys, key)
 	}
+	sort.Sort(version.Collection(keys))
 
-	sort.Sort(version.Collection(versions))
+	// set the current versions
+	if itemLen := len(keys); itemLen >= 1 {
```

Comment:
then we would need to get the len again to do the end of slice element we get in the following line. Can be simplified with the reverse sort. need to look at it all in that context tho.